router/reqlog/reqlog.go

```diff
@@ -0,0 +1,102 @@
+package reqlog
+
+import (
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+)
+
+type Log struct {
+	ID         string
+	Method     string
+	Path       string
+	Host       string
+	RemoteAddr string
+	Backend    string
+	Start      time.Time
+	Finish     time.Time
+	StatusCode int
+	BodyBytes  uint64
+
+	// heroku-isms:
+
+	Level string
+	Code  string
+}
+
+func (rl *Log) WriteTo(logger log15.Logger) {
+	logger.Info("request",
+		"method", rl.Method,
+		"path", rl.Path,
+		"host", rl.Host,
+		"fwd", rl.RemoteAddr,
+		"backend", rl.Backend,
+		"service", durationMS(rl.Start, rl.Finish),
+		"status", rl.StatusCode,
+		"bytes.body", rl.BodyBytes,
+		"request_id", rl.ID,
+	)
```

Comment:
oh, I meant to mention on here last night… This is currently logging synchronously, which is a problem for a high performance HTTP router. Basically we want to router to give up connections and return them to the pool as fast as possible, but we might be introducing a substantial delay by blocking on a write to stdout. Ideally we would throw these messages into a queue of some kind so they can just be logged asynchronously by another goroutine.

@titanous was mentioning log15's [Handlers](https://godoc.org/gopkg.in/inconshreveable/log15.v2#hdr-Handlers) as a possible solution for this. I haven't spent enough time looking at that to know if it's appropriate, but if there are other examples of this elsewhere in Flynn, please lmk!